javalib/src/main/scala/java/util/Arrays.scala

```diff
@@ -317,85 +418,179 @@ object Arrays {
       throw new ArrayIndexOutOfBoundsException
   }
 
-  def hashCode(a: Array[Boolean]): Int = {
-    if (a == null) 0
-    else a.foldLeft(1)((acc, x) => 31*acc + new java.lang.Boolean(x).hashCode)
-  }
+  def asList[T](a: Array[T]): List[T] = {
+    new AbstractList[T] with RandomAccess {
+      def size(): Int =
+        a.length
 
-  def hashCode(a: Array[Char]): Int = {
-    if (a == null) 0
-    else a.foldLeft(1)((acc, x) => 31*acc + new java.lang.Character(x).hashCode)
-  }
+      def get(index: Int): T =
+        a(index)
 
-  def hashCode(a: Array[Byte]): Int = {
-    if (a == null) 0
-    else a.foldLeft(1)((acc, x) => 31*acc + new java.lang.Byte(x).hashCode)
+      override def set(index: Int, element: T): T = {
+        val ret = a(index)
+        a(index) = element
+        ret
+      }
+    }
   }
 
-  def hashCode(a: Array[Short]): Int = {
-    if (a == null) 0
-    else a.foldLeft(1)((acc, x) => 31*acc + new java.lang.Short(x).hashCode)
-  }
+  def hashCode(a: Array[Long]): Int =
+    hashCodeImpl[Long](a, x => new java.lang.Long(x).hashCode)
 
-  def hashCode(a: Array[Int]): Int = {
-    if (a == null) 0
-    else a.foldLeft(1)((acc, x) => 31*acc + new java.lang.Integer(x).hashCode)
-  }
+  def hashCode(a: Array[Int]): Int =
+    hashCodeImpl[Int](a, x => new java.lang.Integer(x).hashCode)
 
-  def hashCode(a: Array[Long]): Int = {
-    if (a == null) 0
-    else a.foldLeft(1)((acc, x) => 31*acc + new java.lang.Long(x).hashCode)
-  }
+  def hashCode(a: Array[Short]): Int =
+    hashCodeImpl[Short](a, x => new java.lang.Short(x).hashCode)
+
+  def hashCode(a: Array[Char]): Int =
+    hashCodeImpl[Char](a, x => new java.lang.Character(x).hashCode)
+
+  def hashCode(a: Array[Byte]): Int =
+    hashCodeImpl[Byte](a, x =>  new java.lang.Byte(x).hashCode)
+
+  def hashCode(a: Array[Boolean]): Int =
+    hashCodeImpl[Boolean](a, x => new java.lang.Boolean(x).hashCode)
+
+  def hashCode(a: Array[Float]): Int =
+    hashCodeImpl[Float](a, x => new java.lang.Float(x).hashCode)
+
+  def hashCode(a: Array[Double]): Int =
+    hashCodeImpl[Double](a, x => new java.lang.Double(x).hashCode)
 
-  def hashCode(a: Array[Float]): Int = {
+  def hashCode(a: Array[AnyRef]): Int =
+    hashCodeImpl[AnyRef](a, x => if (x == null) 0 else x.hashCode)
+
+  @inline
+  private def hashCodeImpl[@specialized T](a: Array[T], elementHashCode: T => Int): Int = {
     if (a == null) 0
-    else a.foldLeft(1)((acc, x) => 31*acc + new java.lang.Float(x).hashCode)
+    else a.foldLeft(1)((acc, x) => 31*acc + elementHashCode(x))
```

Comment:
No need for `elementHashCode`. You can use `x.asInstanceOf[AnyRef].hashCode` instead.